File: tests/regression/commerce_http.php
<?php
/** Loopback HTTP server: real route files, JWT verification, DB identities/roles and middleware.
 * Endpoint bodies are read-only sentinels to avoid exercising unrelated business providers.
 */
require __DIR__.'/commerce_bootstrap.php';
use think\facade\Db;
function request(){return app()->request;}
function filter_str($value){return $value;}
$testApp->setNamespace('app\\adminapi');
$testApp->http->name('adminapi');
$testApp->config->set(['url_route_must'=>true,'route_complete_match'=>true,'url_convert'=>false],'route');
$testApp->config->set(['token_name'=>'Authori-zation','header'=>[]],'cookie');
$testApp->config->set(['default'=>'file','channels'=>['file'=>['type'=>'File','path'=>sys_get_temp_dir().'/crmeb-hardening-http-log/']]],'log');
$testApp->env->set(['app'=>['app_key'=>'isolated-http-jwt-key-not-a-production-secret']]);
$testApp->instance('request',app\Request::__make($testApp));
$testApp->bind(think\exception\Handle::class,app\adminapi\AdminApiExceptionHandle::class);
$testApp->instance('json',new class {
    public function success($data=[]){return think\Response::create(['status'=>200,'data'=>$data],'json');}
    public function make($status,$msg,$data=[]){return think\Response::create(['status'=>$status,'msg'=>$msg,'data'=>$data],'json');}
    public function fail($msg,$data=[]){return $this->make(400,$msg,$data);}
});
class HttpControllerSentinel {
    private $kind; public function __construct($kind){$this->kind=$kind;}
    public function index(){return $this->body();}
    public function lst(){return $this->body();}
    public function refund_update($id){return $this->body();}
    public function body(){return think\Response::create(['status'=>200,'kind'=>$this->kind,'admin_id'=>request()->adminId()],'json');}
}
foreach(['app\\adminapi\\controller\\v1\\finance\\UserRecharge'=>'finance','app\\adminapi\\controller\\v1\\order\\StoreOrder'=>'order','app\\adminapi\\controller\\v1\\user\\User'=>'user']as $class=>$kind)$testApp->bind($class,function()use($kind){return new HttpControllerSentinel($kind);});
if(PHP_SAPI==='cli'&&($argv[1]??'')==='seed'){
    hardeningSchema();
    foreach(['system_menus','system_role','system_admin']as $table)Db::name($table)->where('1=1')->delete();
    foreach(['finance/recharge','order/list','user/user']as $i=>$path)Db::name('system_menus')->insert(['id'=>$i+1,'auth_type'=>2,'api_url'=>$path,'methods'=>'GET']);
    Db::name('system_menus')->insert(['id'=>4,'auth_type'=>2,'api_url'=>'finance/recharge/<id>','methods'=>'PUT']);
    Db::name('system_role')->insert(['id'=>1,'role_name'=>'read-fixtures','rules'=>'1,2,3','status'=>1]);
    Db::name('system_role')->insert(['id'=>2,'role_name'=>'refund-fixtures','rules'=>'4','status'=>1]);
    foreach([[1,1,''],[2,1,'1'],[3,1,'2'],[4,0,'']]as [$id,$level,$roles])Db::name('system_admin')->insert(['id'=>$id,'account'=>'synthetic-'.$id,'pwd'=>'synthetic-password-hash','level'=>$level,'roles'=>$roles,'status'=>1,'is_del'=>0]);
    crmeb\services\CacheService::clearAll();
    $tokens=[];
    foreach([1,2,3,4]as $id)$tokens[$id]=(new crmeb\utils\JwtAuth)->createToken($id,'admin',['pwd'=>md5('synthetic-password-hash')])['token'];
    $tokens['consumer']=(new crmeb\utils\JwtAuth)->createToken(1,'api')['token'];
    echo json_encode($tokens);exit;
}
try {
    $response=$testApp->route->dispatch(request(),function(){
        foreach(['finance','order','user']as $file)require dirname(__DIR__,2).'/crmeb/app/adminapi/route/'.$file.'.php';
        // Known route with missing metadata must be denied, including CRUD.
        think\facade\Route::get('crud/unregistered',function(){return think\Response::create(['status'=>200],'json');})
            ->middleware([app\adminapi\middleware\AdminAuthTokenMiddleware::class,app\adminapi\middleware\AdminCheckRoleMiddleware::class]);
    });
    $response->send();
}catch(crmeb\exceptions\AuthException $e){http_response_code($e->getCode()===401?401:403);header('Content-Type: application/json');echo json_encode(['status'=>http_response_code(),'msg'=>$e->getMessage()]);}
catch(think\exception\RouteNotFoundException $e){http_response_code(404);echo '{}';}
catch(Throwable $e){http_response_code(500);header('Content-Type: application/json');echo json_encode(['error'=>get_class($e),'message'=>$e->getMessage()]);}

File: tests/regression/commerce_http.py
"""Real loopback HTTP authorization tests. Requires isolated MySQL and PHP_BINARY/PHPRC."""
import json, os, pathlib, socket, subprocess, tempfile, time, urllib.request, urllib.error
root=pathlib.Path(__file__).resolve().parents[2]
php=os.environ.get('PHP_BINARY',str(root/'.build/php74/php.exe'))
env=dict(os.environ)
assert env.get('CRMEB_AUDIT_DATABASE')=='crmeb_hardening_test'
tokens=json.loads(subprocess.check_output([php,str(root/'tests/regression/commerce_http.php'),'seed'],env=env,cwd=root))
with socket.socket() as sock:
    sock.bind(('127.0.0.1',0));port=sock.getsockname()[1]
with tempfile.TemporaryFile() as log:
    server=subprocess.Popen([php,'-S',f'127.0.0.1:{port}',str(root/'tests/regression/commerce_http.php')],cwd=root,env=env,stdout=log,stderr=log)
    try:
        deadline=time.monotonic()+10
        while True:
            try:
                with socket.create_connection(('127.0.0.1',port),timeout=.2):break
            except OSError:
                if time.monotonic()>deadline:raise RuntimeError('HTTP fixture server unavailable')
                time.sleep(.05)
        def check(path,identity,expected,method='GET'):
            headers={'Accept':'application/json'}
            if identity is not None:headers['Authori-zation']='Bearer '+tokens[str(identity)]
            req=urllib.request.Request(f'http://127.0.0.1:{port}/'+path,headers=headers,method=method)
            try:
                with urllib.request.urlopen(req,timeout=10) as res:status=res.status;body=res.read().decode()
            except urllib.error.HTTPError as res:status=res.code;body=res.read().decode()
            actual=json.loads(body).get('status',status)
            assert actual==expected,(identity,method,path,status,body)
            print(f'PASS: HTTP {method} {path}, identity={identity}, HTTP={status}, business={actual}')
        for path in ['finance/recharge','order/list','user/user']:
            check(path,None,401);check(path,1,400);check(path,2,200);check(path,3,400);check(path,4,200);check(path,'consumer',401)
        check('finance/recharge/1',2,400,'PUT');check('finance/recharge/1',3,200,'PUT')
        check('crud/unregistered',1,400);check('crud/unregistered',2,400)
        check('not-a-route',4,404)
        print('HTTP authorization matrix passed.')
    finally:
        server.terminate();server.wait(timeout=10)
        log.seek(0)
        (root/'.build/commerce-hardening-20260914/http-test.log').write_bytes(log.read())
